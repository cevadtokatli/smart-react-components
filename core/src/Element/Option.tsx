import styled from 'styled-components'
import applyStyledProps, { StyledProps, shouldForwardProp } from '../styled-props'

export default styled.option.withConfig({ shouldForwardProp })<StyledProps>(applyStyledProps)
