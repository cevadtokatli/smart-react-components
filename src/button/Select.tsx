import React from "react"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, ValueMix, SetValue, JSXChild} from "../props"
import DV from "../default-value"
import useSelectMethods from "../form/UseSelectMethods"
import {FixedBoxOnRef} from "../fixed-box/FixedBox"
import ButtonElement from "./Element"
import Loading from "./Loading"
import WaveEffect from "../wave-effect/WaveEffect"
import FixedBox from "../fixed-box/FixedBox"
import SelectFixedBoxElement from "../form/SelectFixedBoxElement"
import Div from "../element/Div"
import Input from "../form/Input"

export interface Props extends SizeProps {
    buttonProps?: JSXElementProps
    boxProps?: JSXElementProps
    children: JSX.Element[]
    type?: string
    fill?: boolean
    shape?: string
    block?: boolean
    waveEffect?: string
    active: ValueMix
    setActive: SetValue<ValueMix>
    placeholder?: JSXChild
    hover?: boolean
    optionType?: string
    boxShadow?: boolean
    minWidth?: number
    transitionClassName?: string
    transitionType?: string
    transitionDuration?: number
    showAnimation?: boolean
    hideAnimation?: boolean
    loading?: JSX.Element
    loadingStatus?: boolean
    loadingTransitionClassName?: string
    loadingTransitionType?: string
    loadingTransitionDuration?: number 
    loadingTransitionShowAnimation?: boolean
    loadingTransitionHideAnimation?: boolean
    search?: boolean
    searchInputProps?: JSXElementProps
}

const ButtonSelect: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,buttonProps=DV.JSX_ELEMENT_PROPS,boxProps=DV.JSX_ELEMENT_PROPS,children,type="primary",fill,shape="default",block,waveEffect="light",active,setActive,placeholder,hover=true,optionType="primary",boxShadow,minWidth=300,transitionClassName,transitionType,transitionDuration,showAnimation=true,hideAnimation=true,loading,loadingStatus,loadingTransitionClassName="src-button-loading",loadingTransitionType="ease-in-out",loadingTransitionDuration=300,loadingTransitionShowAnimation=true,loadingTransitionHideAnimation=true,search,searchInputProps=DV.JSX_ELEMENT_PROPS}) => {
    const {itemList,searchText,setSearchText} = useSelectMethods({size,sizeSm,sizeMd,sizeLg,sizeXl,box:true,placeholder,optionType,children,fill,shape,hover,disabled:buttonProps.disabled,active,setActive,closeBox:(() => setStatus(false)),search})
    const [status, setStatus] = React.useState<boolean>(false)
    const fixedBoxOnRef = React.useRef<FixedBoxOnRef>(null)

    React.useEffect(() => {
        if(fixedBoxOnRef.current)
            fixedBoxOnRef.current.setPosition()
    }, [itemList])

    const ButtonEl = (
        <ButtonElement tabIndex={0} {...buttonProps} as="div" size$={size} size$Sm={sizeSm} size$Md={sizeMd} size$Lg={sizeLg} size$Xl={sizeXl} type$={type} fill$={fill} shape={shape} block={block} loading$={loading ? true : false} data-src-placeholder-style={itemList.isPlaceholder} data-src-active={status}>
            {loading && <Loading element={loading} status={loadingStatus} transitionClassName={loadingTransitionClassName} transitionDuration={loadingTransitionDuration} transitionType={loadingTransitionType} transitionShowAnimation={loadingTransitionShowAnimation} transitionHideAnimation={loadingTransitionHideAnimation} />}
            {itemList.active}
        </ButtonElement>
    )

    const FragmentEl = waveEffect ? <WaveEffect type={waveEffect}>{ButtonEl}</WaveEffect> : ButtonEl

    return (
        <FixedBox status={status} setStatus={setStatus} minWidth={minWidth} boxShadow={boxShadow} transitionClassName={transitionClassName} transitionType={transitionType} transitionDuration={transitionDuration} showAnimation={showAnimation} hideAnimation={hideAnimation} onRef={args => fixedBoxOnRef.current = args}>
            {FragmentEl}
            <SelectFixedBoxElement {...boxProps} size={size} sizeSm={sizeSm} sizeMd={sizeMd} sizeLg={sizeLg} sizeXl={sizeXl} type={type} shape={shape}>
                <Div height="inherit" minHeight="inherit" maxHeight="inherit">
                    {search && <section><Input {...searchInputProps} value={searchText} setValue={setSearchText} /></section>}
                    {itemList.value}
                </Div>
            </SelectFixedBoxElement>
        </FixedBox>
    )
}

export default ButtonSelect