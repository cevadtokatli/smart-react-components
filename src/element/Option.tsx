import styled from "styled-components"
import {Props, attrs, css} from "./"

export default styled.option.attrs<Props>(attrs)<Props>(css)