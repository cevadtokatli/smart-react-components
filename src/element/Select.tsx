import styled from "styled-components"
import {Props, attrs, css} from "./"

export default styled.select.attrs<Props>(attrs)<Props>(css)