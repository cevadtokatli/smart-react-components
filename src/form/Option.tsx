import React from "react"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChild, Value, ValueMix, SetValue} from "../props"
import DV from "../default-value"
import OptionElement from "./OptionElement"

export interface Props extends SizeProps {
    elementProps?: JSXElementProps
    children: JSXChild
    value: Value
    searchValue?: string
    disabled?: boolean
    box?: boolean
    type?: string
    fill?: boolean
    shape?: string
    hover?: boolean
    space?: boolean
    active?: ValueMix
    setActive?: SetValue<ValueMix>
}

const Option: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,elementProps=DV.JSX_ELEMENT_PROPS,children,value,disabled,box,type="primary",fill,shape="default",hover,active,setActive}) => {
    /**
     * Returns active$ status of the element. 
     */
    const getActive$ = (): boolean => {
        if(value == null)
            return false
        return Array.isArray(active) ? active.indexOf(value) > -1 : value == active
    }

    const [active$, setActive$] = React.useState<boolean>(() => getActive$())
    
    React.useEffect(() => setActive$(getActive$()), [value, active])

    /**
     * Sets value if element is not disabled.
     */
    const set = (): void => {
        if(disabled || (value == null && Array.isArray(active)))
            return

        if(Array.isArray(active)) {
            if(active.indexOf(value) == -1)
                setActive([...active,value])
            else
                setActive(active.filter(i => i != value))
        } else if(value != active)
            setActive(value)
    }

    return <OptionElement {...elementProps} className={`${active$ ? "src-option-active" : ""}`} size={size} sizeSm={sizeSm} sizeMd={sizeMd} sizeLg={sizeLg} sizeXl={sizeXl} box={box} type={type} fill$={fill} shape={shape} hover={hover} disabled={disabled} value={value} onClick={set}>{children}</OptionElement>
}

export default Option