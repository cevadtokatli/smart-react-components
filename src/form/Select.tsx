import React from "react"
import Props from "./SelectProps"
import DV from "../default-value"
import UseSelectMethods from "./UseSelectMethods"
import UseInputOutput from "./UseInputOutput"
import SelectElement from "./InputElement"
import Input from "./Input"

const Select: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,containerProps=DV.JSX_ELEMENT_PROPS,selectProps=DV.JSX_ELEMENT_PROPS,labelProps=DV.JSX_ELEMENT_PROPS,label,children,optionType="primary",active,setActive,type,fill,border,shape="default",hover=true,error,leftAddon,leftAddonProps=DV.JSX_ELEMENT_PROPS,leftAddonAbsolute,rightAddon,rightAddonProps=DV.JSX_ELEMENT_PROPS,rightAddonAbsolute,search,searchInputProps=DV.JSX_ELEMENT_PROPS}) => {
    const {itemList,searchText,setSearchText} = UseSelectMethods({children,optionType,size,sizeSm,sizeMd,sizeLg,sizeXl,fill,shape,hover,disabled:selectProps.disabled,active,setActive,search}) 
       
    return UseInputOutput({
        element: (
            <SelectElement {...selectProps} size$={size} size$Sm={sizeSm} size$Md={sizeMd} size$Lg={sizeLg} size$Xl={sizeXl} type$={type} fill$={fill} border$={border} shape={shape} leftAddon={leftAddon ? true : false} leftAddonAbsolute={leftAddonAbsolute} rightAddon={rightAddon ? true : false} rightAddonAbsolute={rightAddonAbsolute} inputElement={false}>
                {search && <Input {...searchInputProps} value={searchText} setValue={setSearchText} />}
                {itemList.value}
            </SelectElement>
        ),
        containerProps,inputProps:selectProps,labelProps,label,shape,leftAddon,leftAddonProps,leftAddonAbsolute,rightAddon,rightAddonProps,rightAddonAbsolute,error
    })
}

export {default as Props} from "./SelectProps"
export default Select