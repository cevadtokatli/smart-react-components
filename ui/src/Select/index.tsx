import Div from '@smart-react-components/core/Element/Div'
import { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import { ContentElement, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import FormBlockLabel from '../components/FormBlockLabel'
import useInputAddons from '../hooks/useInputAddons'
import { FormValue } from '../types'
import SelectElement from './SelectElement'
import HiddenInput from '../components/HiddenInput'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active: FormValue | FormValue[]
  children: JSX.Element | JSX.Element[]
  hasBorder?: boolean
  hasHover?: boolean
  hasWaveEffect?: boolean
  isBlock?: boolean
  isDisabled?: boolean
  isRequired?: boolean
  isOutline?: boolean
  isSoft?: boolean
  label?: ContentElement
  leftAddon?: ContentElement
  palette?: PaletteProp
  rightAddon?: ContentElement
  shape?: ShapeProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette?: PaletteProp
}

const Select = React.forwardRef<HTMLElement, Props>((props, forwardRef) => {
  const { leftAddon, rightAddon } = useInputAddons({
    hasBorder: props.hasBorder,
    isDisabled: props.isDisabled,
    isFocused: false,
    isOutline: props.isOutline,
    isSoft: props.isSoft,
    leftAddon: props.leftAddon,
    palette: props.palette,
    rightAddon: props.rightAddon,
    shape: props.shape,
    size: props.size,
    sizeSm: props.sizeSm,
    sizeMd: props.sizeMd,
    sizeLg: props.sizeLg,
    sizeXl: props.sizeXl,
  })

  const getInputValue = () => (
    (!Array.isArray(props.active) && typeof props.active !== 'undefined' && props.active !== null)
    || (Array.isArray(props.active) && props.active.length > 0)
  )
    ? String(props.active)
    : ''

  return (
    <FormBlockLabel
      formSize={props.size}
      formSizeSm={props.sizeSm}
      formSizeMd={props.sizeMd}
      formSizeLg={props.sizeLg}
      formSizeXl={props.sizeXl}
      isBlock={props.isBlock}
      isDisabled={props.isDisabled}
    >
      { props.label && <span>{props.label}</span> }
      <Div
        display="flex"
        flex="1 1 auto"
      >
        { leftAddon && leftAddon }
        <SelectElement
          hasBorder={props.hasBorder}
          hasLeftAddon={!!leftAddon}
          hasRightAddon={!!rightAddon}
          hasSeparatedLeftAddon={leftAddon?.props?.isSeparated}
          hasSeparatedRightAddon={rightAddon?.props?.isSeparated}
          inputSize={props.size}
          inputSizeSm={props.sizeSm}
          inputSizeMd={props.sizeMd}
          inputSizeLg={props.sizeLg}
          inputSizeXl={props.sizeXl}
          isDisabled={props.isDisabled}
          isFocused={false}
          isOutline={props.isOutline}
          isSoft={props.isSoft}
          palette={props.palette}
          shape={props.shape}
        >
          <HiddenInput
            onChange={() => {}}
            ref={forwardRef}
            value={getInputValue()}
            {...(props.isDisabled && { disabled: true })}
            {...(props.isRequired && { required: true })}
          />
          { (Array.isArray(props.children) ? props.children : [props.children]).map((item, idx) => React.cloneElement(item, {
            key: item.key ?? idx,
            active: props.active,
            hasHover: props.hasHover,
            hasWaveEffect: props.hasWaveEffect,
            isEmbedded: true,
            isOutline: props.isOutline,
            isSoft: props.isSoft,
            palette: props.palette,
            setActive: props.setActive,
            waveEffectPalette: props.waveEffectPalette,
            ...(props.isDisabled && { isDisabled: true }),
          })) }
        </SelectElement>
        { rightAddon && rightAddon }
      </Div>
    </FormBlockLabel>
  )
})

Select.defaultProps = {
  hasBorder: true,
  hasHover: true,
  hasWaveEffect: true,
  isBlock: true,
  isOutline: true,
  palette: 'primary',
  shape: 'rectangle',
  size: 'medium',
  waveEffectPalette: 'light',
}

export default Select
