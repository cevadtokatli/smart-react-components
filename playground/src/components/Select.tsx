import Select from "@smart-react-components/ui/Select"
import Option from "@smart-react-components/ui/Select/Option"
import OptionGroup from "@smart-react-components/ui/Select/OptionGroup"
import React from 'react'

export const Palette = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <form>
      <Select active={active} setActive={setActive} isRequired>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <button>butotn</button>
      </form>
      <Select palette="secondary" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="success" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="danger" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="warning" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="info" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="dynamic" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="!dynamic" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}

export const Soft = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <Select active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="secondary" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="success" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="danger" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="warning" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="info" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="dynamic" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="!dynamic" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}

export const Outline = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <Select active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="secondary" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="success" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="danger" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="warning" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="info" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="dynamic" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="!dynamic" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}

export const SoftOutline = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <Select active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="secondary" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="success" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="danger" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="warning" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="info" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="dynamic" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select palette="!dynamic" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}

export const Size = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <Select active={active} setActive={setActive} size="small" label="Small">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select active={active} setActive={setActive} label="Medium">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select active={active} setActive={setActive} size="large" label="Large">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}

export const Shape = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <Select active={active} setActive={setActive} label="Rectangle">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
      <Select active={active} setActive={setActive} label="Rounded" shape="rounded">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    </>
  )
}
