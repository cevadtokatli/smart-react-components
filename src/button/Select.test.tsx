import React from "react"
import renderTheme from "../test/renderTheme"
import mountTheme from "../test/mountTheme"
import Option from "../form/Option"
import Select from "./Select"

describe("<Select />", () => {
    let children
    let props

    beforeAll(() => {
        children = [
            <Option key="0" value="value-1">Value-1</Option>,
            <Option key="1" value="value-2">Value-2</Option>,
            <Option key="2" value="value-3">Value-3</Option>
        ]
    })

    beforeEach(() => {
        props = {
            active: "value-2",
            setActive: jest.fn(),
            buttonProps: {
                id: "button",
                onClick: e => e.target.click()
            },
            showAnimation: false,
            hideAnimation: false
        }
    })

    it("should match snapshot", () => {
        const wrapper = renderTheme(<Select {...props}>{children}</Select>)
        expect(wrapper).toMatchSnapshot()
    })

    it("should call setActive", () => {
        const wrapper = mountTheme(<Select {...props}>{children}</Select>)
        wrapper.find("#button").hostNodes().simulate("click")
        wrapper.find(Option).at(0).simulate("click")
        expect(props.setActive).toHaveBeenCalled()
    })

    it("should render placeholder", () => {
        props.active = null
        const wrapper = mountTheme(<Select {...props} placeholder={<span id="placeholder" />}>{children}</Select>)
        expect(wrapper.find("#placeholder").exists()).toBe(true)
    })

    it("should render loading element", () => {
        const wrapper = mountTheme(<Select {...props} loading={<span id="loading" />} loadingStatus={true}>{children}</Select>)
        expect(wrapper.find("#loading").exists()).toBe(true)
    })

    it("should render search input", () => {
        const wrapper = mountTheme(<Select {...props} search={true} searchInputProps={{id:"search-input"}}>{children}</Select>)
        wrapper.find("#button").hostNodes().simulate("click")
        expect(wrapper.find("#search-input").exists()).toBe(true)
    })
})