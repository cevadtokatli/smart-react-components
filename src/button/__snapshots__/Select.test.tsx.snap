// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should match snapshot 1`] = `
<div
  class="sc-fzpans sc-fzplWN cXylyC"
  data-src-active="false"
  data-src-placeholder-style="false"
  data-src-wave-effect="0"
  id="button"
  shape="default"
  tabindex="0"
>
  <div
    class="sc-AxjAm bSfLnR"
    display="flex"
  >
    <div
      class="sc-AxjAm jfvPlI"
    >
      Value-2
    </div>
    <svg
      class="sc-AxhUy sc-fzozJi iXmFV"
      viewBox="5 4 14 14"
    >
      <path
        d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
      />
    </svg>
  </div>
</div>
`;
